torch/csrc/StorageSharing.cpp

```diff
@@ -312,36 +310,36 @@ static PyObject* THPStorage_shareCuda(PyObject* self, PyObject* noargs) {
   Py_INCREF(Py_None);
   THPObjectPtr _event_sync_required(Py_None);
   Py_INCREF(Py_None);
+
   if (storage.data()) {
-    // NOLINTNEXTLINE(cppcoreguidelines-init-variables)
-    auto shandle =
-        c10::cuda::CUDACachingAllocator::shareIpcHandle(storage.mutable_data());
-    _handle = PyBytes_FromStringAndSize(
-        shandle.handle.c_str(), (Py_ssize_t)shandle.handle.size());
-    _offset_bytes = PyLong_FromSsize_t((Py_ssize_t)shandle.offset);
-
-    // Put Storage Data behind new ref counting context
-    // See Note [CUDA IPC Refcounting implementation explained]
-    at::DataPtr sent_data_ptr = torch::GetNewRefCountedSentData(
-        storage.mutable_data(), storage.device());
-    auto old_data_ptr = storage.set_data_ptr(std::move(sent_data_ptr));
-    auto sent_data =
-        static_cast<torch::CudaIPCSentData*>(storage.data_ptr().get_context());
-    sent_data->set_original_ptr(std::move(old_data_ptr));
-    _ref_counter = PyBytes_FromString((sent_data->handle()).c_str());
-    _ref_counter_offset = THPUtils_packUInt64(sent_data->offset());
-
-    // NOLINTNEXTLINE(cppcoreguidelines-init-variables)
-    cudaIpcEventHandle_t ipc_event_handle;
-
-    if (sent_data->event_sync_required_) {
-      C10_CUDA_CHECK(
-          cudaIpcGetEventHandle(&ipc_event_handle, sent_data->event_));
-    }
+    size_t ipc_memory_handle_size;
+    size_t ipc_event_handle_size;
+    at::globalContext().getAcceleratorHooksInterface().getIpcHandleSize(
+        ipc_memory_handle_size, ipc_event_handle_size);
+
+    ptrdiff_t offset_bytes;
+    auto new_memory_handle = std::make_unique<char[]>(ipc_memory_handle_size);
+    auto new_event_handle = std::make_unique<char[]>(ipc_event_handle_size);
+    auto new_ref_counter = std::make_unique<char[]>(sizeof(std::string));
```

Comment:
This seems like a pessimization; previously we copied straight from shareIpcHandle into the Python byte string; now you do an intermediate copy through these dynamic allocations

Reply:
@ezyang  Modified